src/scripting_api/app.js

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2020 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { EventDispatcher } from "./event.js";
+import { NotSupportedError } from "./error.js";
+import { PDFObject } from "./pdf_object.js";
+
+class App extends PDFObject {
+  constructor(data) {
+    super(data);
+    this._specialId = true;
+    this._document = data._document;
+    this._setTimeout = setTimeout;
+    this._clearTimeout = clearTimeout;
+    this._setInterval = setInterval;
+    this._clearInterval = clearInterval;
+    this._calculationOrder = data.calculationOrder;
+    this._objects = Object.create(null);
+    this._eventDispatcher = new EventDispatcher(
+      this._document,
+      data.calculationOrder,
+      this._objects
+    );
+  }
+
+  _dispatchEvent(pdfEvent) {
+    this._eventDispatcher.dispatch(pdfEvent);
+  }
+
+  get activeDocs() {
+    return [this._document.wrapped];
+  }
+
+  set activeDocs(_) {
+    throw new NotSupportedError("app.activeDocs");
+  }
+
+  alert(
+    cMsg,
+    nIcon = 0,
+    nType = 0,
+    cTitle = "PDF.js",
+    oDoc = null,
+    oCheckbox = null
+  ) {
+    this._send({ command: "alert", value: cMsg });
+  }
+
+  clearInterval(oInterval) {
+    return this._clearInterval(oInterval);
+  }
+
+  clearTimeOut(oTime) {
+    return this._clearTimeout(oTime);
+  }
+
+  setInterval(cExpr, nMilliseconds = 1000) {
+    return this._setInterval({ cExpr, nMilliseconds });
+  }
+
+  setTimeOut(cExpr, nMilliseconds = 1000) {
+    return this._setTimeout({ cExpr, nMilliseconds });
```

Comment:
Does this work? AFAICT `this._setTimeout` is equal to the default JS `setTimeout` function, but that doesn't take an object, only separate parameters (the same applies for the other timeout/interval methods above).

Reply:
I removed that stuff for now.
But in fact we call a wrapper (see PdfSandbox.jsm at line 122 in m-c patch) to track the ids and to avoid to have a script which will remove an id it hadn't defined before.